x/valset-pref/msg_server_test.go

```diff
@@ -292,3 +292,123 @@ func (suite *KeeperTestSuite) TestUnDelegateFromValidatorSet() {
+			newPreferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(6, 1),
+				},
+			},
+			coinToStake: sdk.NewCoin("stake", sdk.NewInt(20_000_000)),
+			expectPass:  true,
+		},
+		{
+			name:      "update the same validator set as above",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			newPreferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			coinToStake: sdk.NewCoin("stake", sdk.NewInt(20_000_000)),
+			expectPass:  true,
+		},
+		{
+			name:      "update existing Validator with new validators",
+			delegator: sdk.AccAddress([]byte("addr2---------------")),
+			newPreferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[3],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[4],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[5],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[6],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[7],
+					Weight:         sdk.NewDecWithPrec(1, 1),
+				},
+				{
+					ValOperAddress: valAddrs[8],
+					Weight:         sdk.NewDecWithPrec(1, 1),
+				},
+			},
+			coinToStake: sdk.NewCoin("stake", sdk.NewInt(50_000_000)),
+			expectPass:  true,
+		},
+	}
+
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+
+			// fund the account that is trying to delegate
+			suite.FundAcc(test.delegator, sdk.Coins{sdk.NewInt64Coin(sdk.DefaultBondDenom, 100_000_000)})
+
+			// setup message server
+			msgServer := valPref.NewMsgServerImpl(suite.App.ValidatorSetPreferenceKeeper)
+			c := sdk.WrapSDKContext(suite.Ctx)
+
+			// call the create validator set preference
+			preferences := suite.PrepareDelegateToValidatorSet()
+
+			// call the create validator set preference
+			_, err := msgServer.SetValidatorSetPreference(c, types.NewMsgSetValidatorSetPreference(test.delegator, preferences))
+			suite.Require().NoError(err)
+
+			// call the create validator set preference
+			_, err = msgServer.DelegateToValidatorSet(c, types.NewMsgDelegateToValidatorSet(test.delegator, test.coinToStake))
+			suite.Require().NoError(err)
+
+			_, err = msgServer.RedelegateValidatorSet(c, types.NewMsgRedelegateValidatorSet(test.delegator, test.newPreferences))
+			if test.expectPass {
+				// TODO: check if the validator have recieved the correct amount of tokens
```

Comment:
This should be a part of this PR

Reply:
Only way to correctly check redelegation is to mature it, but for that we need {matureTime, SrcVal, DestVal}
This will mean we have to make our proto response return {matureTime, ValSrc, ValDest} for every single redelegation. Am open to discussing this in our next chain dev call or valset-pref call